Translate Synchronizing-with-effects to Spanish (#673)
reactjs/es.react.dev: src/content/learn/synchronizing-with-effects.md

</Sandpack>

Verás tres mensajes al principio: `Programar mensaje "a"`, `Cancelar mensaje "a"`, y `Programar mensaje "a"` de nuevo. Tres segundos después también aparecerá un mensaje que dice `a`. Como aprendiste anteriormente, el programar/cancelar extra se debe a que React vuelve a montar el componente en desarrollo para verificar que has implementado correctamente la limpieza.

> Ahora edita el campo de texto para que diga `abc`. Si lo hace lo suficientemente rápido, verá `Programar mensaje "ab"` seguido inmediatamente de `Cancelar mensaje "ab"` y `Programar mensaje "abc"`. **React siempre limpia el Efecto del renderizado anterior antes de ejecutar el Efecto de la siguiente renderización** Es porque, incluso si escribe rápidamente en el campo de texto, hay como máximo un temporizador programado a la vez. Edita el campo varias veces y observe la consola para tener una idea de como se limpian los Efectos.

Reviewer: ```suggestion
Ahora edita el campo de texto para que diga `abc`. Si lo haces lo suficientemente rápido, verás `Programar mensaje "ab"` seguido inmediatamente de `Cancelar mensaje "ab"` y `Programar mensaje "abc"`. **React siempre limpia el Efecto del renderizado anterior antes de ejecutar el Efecto del siguiente renderizdo.** Es por eso que incluso si escribes rápidamente en el campo de texto, hay como máximo un temporizador programado a la vez. Edita el campo varias veces y observa la consola para tener una idea de cómo se limpian los Efectos.
```

Author: Hola @carburo! Gracias por las sugerencias (algunas hasta me dan verguenza haberlas cometido, tuve que haber revisado mejor) acepté todas menos esta que escribiste renderizdo

Reviewer: Corregido, ¡gracias! :sweat_smile:

Escribe algo dentro del campo e inmediatamente presiona "Desmontar componente". Observa como al desmontar se limpia el Efecto del último renderizado. En este caso, se elimina el último tiempo de espera antes de que tenga la oportunidad de ejecutarse.

Por último, edita el componente de arriba y comenta la función de limpieza para que los tiempos de espera no se cancelen. Intenta escribir `abcde` rápidamente. ¿Qué esperas que suceda en tres segundos? ¿El `console.log(text)` dentro del tiempo de espera imprimirá el `text` *más reciente* y producirá cinco mensajes `abcde`? ¡Inténtalo para comprobar tu intuición!

> Tres segundos después, deberías ver una secuencia de mensajes (`a`, `ab`, `abc`, `abcd`, and `abcde`) en vez de cinco mensajes `abcde`. **Cada Efecto "captura" el valor de `text` desde su correspondiente renderizado.**  No importa que el estado `text` haya cambiado: al renderizar un Efecto con `text = 'ab'` siempre verá `'ab'`. En otras palabras, los Efectos de cada renderizado están aislados entre sí. Si tienes curiosidad acerca de como funciona esto puedes leer acerca de [clausuras](https://developer.mozilla.org/es/docs/Web/JavaScript/Closures).

Reviewer: ```suggestion
Tres segundos después, deberías ver una secuencia de mensajes (`a`, `ab`, `abc`, `abcd`, and `abcde`) en vez de cinco mensajes `abcde`. **Cada Efecto "captura" el valor de `text` desde su correspondiente renderizado.**  No importa que el estado `text` haya cambiado: un Efecto del renderizado hecho con `text = 'ab'` siempre verá `'ab'`. En otras palabras, los Efectos de cada renderizado están aislados entre sí. Si tienes curiosidad acerca de cómo funciona esto puedes leer acerca de [clausuras](https://developer.mozilla.org/es/docs/Web/JavaScript/Closures).
```

    return () => connection.disconnect();
  }, [roomId]);

  return <h1>¡Bienvenido a {roomId}!</h1>;
}
```

> Veamos que es lo que sucede exactamente a medida que el usuario navega por la aplicación.

Reviewer: ```suggestion
Veamos qué es lo que sucede exactamente a medida que el usuario navega por la aplicación.
```

#### Renderizado inicial {/*initial-render*/}

El usuario visita `<ChatRoom roomId="general" />`. Vamos a [sustituir mentalmente](/learn/state-as-a-snapshot#rendering-takes-a-snapshot-in-time) `roomId` con `'general'`:

```js
  // JSX para el primer renderizado (roomId = "general")
  return <h1>¡Bienvenido a general!</h1>;
```

> **El Efecto *también* es una parte de la salida del renderizado.** El efecto del primer renderizado se convierte en:

Reviewer: ```suggestion
**El Efecto *también* es una parte de la salida del renderizado.** El Efecto del primer renderizado se convierte en:
```

Supongamos que `<ChatRoom roomId="general" />` se vuelve a renderizar. La salida JSX es la misma:

```js
  // JSX para el segundo renderizado (roomId = "general")
  return <h1>¡Bienvenido a general!</h1>;
```

React observa que la salida del renderizado no ha cambiado, entonces no actualiza el DOM.

El Efecto del segundo renderizado se ve así:

```js
  // Efecto del segundo renderizado (roomId = "general")
  () => {
    const connection = createConnection('general');
    connection.connect();
    return () => connection.disconnect();
  },
  // Dependencias para el segundo renderizado (roomId = "general")
  ['general']
```

React compara `['general']` del segundo renderizado con `['general']` del primer renderizado. **Como todas las dependencias son iguales, React *ignora* el Efecto del segundo renderizado.** Nunca es llamado.

#### Volver a renderizar con diferentes dependencias {/*re-render-with-different-dependencies*/}

Luego, el usuario visita `<ChatRoom roomId="travel" />`. Esta vez, el componente devuelve un JSX diferente:

```js
  // JSX para el tercer renderizado (roomId = "travel")
  return <h1>¡Bienvenido a travel!</h1>;
```

> React actualiza el DOM para cambiar `"Bienvenido a general"` into `"Bienvenido a  travel"`.

Reviewer: ```suggestion
React actualiza el DOM para cambiar `"Bienvenido a general"` a `"Bienvenido a travel"`.
```

El Efecto del tercer renderizado se ve así:

```js
  // Efecto del tercer renderizado (roomId = "travel")
  () => {
    const connection = createConnection('travel');
    connection.connect();
    return () => connection.disconnect();
  },
  // Dependencias para el tercer renderizado (roomId = "travel")
  ['travel']
```

React compara `['travel']` del tercer renderizado con `['general']` del segundo renderizado. Una dependencia es diferente: `Object.is('travel', 'general')` es `false`. El Efecto no puede ser omitido.

> **Antes de que React pueda aplicar el Efecto de la tercera renderización, necesita limpiar el último efecto que se _ejecutó_.** El Efecto de la segunda renderización fue omitido, por lo que React necesita limpiar el Efecto de la primer renderización. Si te desplazas hacia arriba hasta la primera renderización, verás que su limpieza llama a `disconnect()` en la conexión que se creó con `createConnection('general')`. Esto desconecta la aplicación de la sala de chat `'general'`.

Reviewer: Creo que así se transmite mejor el énfasis que se intenta poner en el original.
```suggestion
**Antes de que React pueda aplicar el Efecto de la tercera renderización, necesita limpiar el último Efecto que _sí_ se ejecutó.** El Efecto de la segunda renderización fue omitido, por lo que React necesita limpiar el Efecto de la primer renderización. Si te desplazas hacia arriba hasta la primera renderización, verás que su limpieza llama a `disconnect()` en la conexión que se creó con `createConnection('general')`. Esto desconecta la aplicación de la sala de chat `'general'`.
```

> Después de eso, React ejecuta el Efecto del tercer renderizado. Que conecta a la sala de chat `'travel'`.

Reviewer: ```suggestion
Después de eso, React ejecuta el Efecto del tercer renderizado que conecta a la sala de chat `'travel'`.
```

#### Desmontar {/*unmount*/}

Finalmente, supongamos que el usuario cambia de página y el componente `ChatRoom` se desmonta. React ejecuta la función de limpieza del último Efecto. El último Efecto fue del tercer renderizado. La limpieza del tercer renderizado destruye la conexión a `createConnection('travel')`. Por lo tanto la aplicación se desconecta de la sala de chat `'travel'`.

#### Comportamiento solo en modo de desarrollo {/*development-only-behaviors*/}

> Cuando el [Modo Estricto](/reference/react/StrictMode) esta activado, React vuelve a montar cada componente después de montarlo por primera vez (el estado y DOM se conservan). Esto [te ayuda a encontrar Efectos que necesitan limpieza](#step-3-add-cleanup-if-needed) y expone errores como las condiciones de carrera tempranas. Adicionalmente, React vuelve a montar los efectos cada vez que guardas el archivo en desarrollo. Ambos comportamientos solo suceden en modo de desarrollo.

Reviewer: ```suggestion
Cuando el [Modo Estricto](/reference/react/StrictMode) está activado, React vuelve a montar cada componente después de montarlo por primera vez (el estado y DOM se conservan). Esto [te ayuda a encontrar Efectos que necesitan limpieza](#step-3-add-cleanup-if-needed) y expone tempranamente errores como las condiciones de carrera. Adicionalmente, React vuelve a montar los Efectos cada vez que guardas el archivo en desarrollo. Ambos comportamientos solo suceden en modo de desarrollo.
```

</DeepDive>

<Recap>

- A diferencia de los eventos, los Efectos son causados por el renderizado en sí mismo en vez de una interacción en particular.
- Los Efectos te permiten sincronizar un componente con sistemas externos (API de terceros, redes, etc).
- Por defecto, los Efectos se ejecutan después de cada renderizado (incluyendo el inicial).
- React omitirá el Efecto si todas sus dependencias tienen los mismos valores que en el anterior renderizado.
- No puedes "elegir" tus dependencias. Son determinadas por el código dentro del Efecto.
- Un array de dependencias vacío (`[]`) corresponde al "montaje" del componente, es decir, cuando se añade en la pantalla.
- En el Modo Estricto, React monta dos veces los componentes (¡Solo en modo de desarrollo!) para poner a prueba tus Efectos.
- Si tu Efecto se rompe debido a que se vuelve a montar, debes implementar una función de limpieza.
- React llamará a tu función de limpieza antes de que el Efecto se ejecute la próxima vez, y también durante el desmontaje.

</Recap>

<Challenges>

#### Enfoca un campo en montaje {/*focus-a-field-on-mount*/}

En este ejemplo, el formulario renderiza un componente `<MyInput />`.

> Usa el método [`focus()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus) del input para hacer que `MyInput` se enfoque automáticamente cuando aparece en pantalla. Ya hay una implementación comentada, pero no funciona correctamente. Descubre porque no funciona y corrígelo. (Si estás familiarizado con el atributo `autoFocus`, pretende que no existe: estamos re-implementando la misma funcionalidad desde cero.)

Reviewer: ```suggestion
Usa el método [`focus()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus) del input para hacer que `MyInput` se enfoque automáticamente cuando aparece en pantalla. Ya hay una implementación comentada, pero no funciona correctamente. Descubre por qué no funciona y corrígelo. (Si estás familiarizado con el atributo `autoFocus`, pretende que no existe: estamos reimplementando la misma funcionalidad desde cero.)
```

</Sandpack>

En desarrollo, React seguirá montando el componente una vez para verificar que hayas implementado bien la limpieza. Por lo tanto, habrá una llamada a `setInterval`, seguida inmediatamente por `clearInterval` y nuevamente `setInterval`. En producción, solo habrá una llamada a `setInterval`. El comportamiento visible para el usuario en ambos casos es el mismo: el contador se incrementa una vez por segundo.

</Solution>

#### Arregla la solicitud dentro de un Efecto {/*fix-fetching-inside-an-effect*/}

> Este componente muestra la biografía de la persona seleccionada. Carga la biografíá llamando a una función asíncrona `fetchBio(person)` al montar y cuando `person` cambia. Esa función asíncrona devuelve una [promesa](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Promise) que eventualmente se resuelve a un string. Cuando la solicitud está hecha, se llama a `setBio` para mostrar el string debajo del elemento select.

Reviewer: ```suggestion
Este componente muestra la biografía de la persona seleccionada. Carga la biografía llamando a una función asíncrona `fetchBio(person)` al montar y cuando `person` cambia. Esa función asíncrona devuelve una [promesa](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Promise) que eventualmente se resuelve a un string. Cuando la solicitud está hecha, se llama a `setBio` para mostrar el string debajo del elemento select.
```

</Sandpack>

> Cada Efecto de renderización tiene su propia variable `ignore`. Inicialmente, la variable `ignore` es `false`. Sin embargo, si un Efecto hace una limpieza (como cuando seleccionas una persona diferente), su variable `ignore` pasa a ser `true`. Each render's Effect has its own `ignore` variable. Initially, the `ignore` variable is set to `false`. However, if an Effect gets cleaned up (such as when you select a different person), its `ignore` variable becomes `true`. Entonces, ahora no importa en qué orden se completen las solicitudes. Solo el Efecto de la última persona tendrá `ignore` configurado como `false`, por lo que llamará a `setBio(result)`. Los Efectos anteriores han sido limpiados, por lo que la comprobación `if (!ignore)` evitará que llamen a `setBio`:

Reviewer: ```suggestion
El Efecto de cada uno de los renderizados tiene su propia variable `ignore`. Inicialmente, la variable `ignore` es `false`. Sin embargo, si un Efecto hace una limpieza (como cuando seleccionas una persona diferente), su variable `ignore` pasa a ser `true`. Entonces, ahora no importa en qué orden se completen las solicitudes. Solo el Efecto de la última persona tendrá `ignore` configurado como `false`, por lo que llamará a `setBio(result)`. Los Efectos anteriores han sido limpiados, por lo que la comprobación `if (!ignore)` evitará que llamen a `setBio`:
```

- Seleccionar a `'Bob'` desencadena `fetchBio('Bob')`

> - Seleccionar a `'Taylor'` desencadena `fetchBio('Taylor')` **y limpia el anterior (Bob's) Efecto**

Reviewer: ```suggestion
- Seleccionar a `'Taylor'` desencadena `fetchBio('Taylor')` **y limpia el anterior Efecto (de Bob)**
```
